File: examples/point_subsets/terraclimate_extract_point_subsets.py
# example script for extracting list of points locations from terraclimate using OPeNDAP in Python 2.7
# For python3 anaconda users, be aware of bug
# https://www.unidata.ucar.edu/mailing_lists/archives/python-users/2017/msg00005.html
#
# Also, an alternative to netCDF4 is xarray, where here's a start:
#     http://climate.nkn.uidaho.edu/TERRACLIMATE/pages/guidance/CODE/terraclimate_opendap_points_python3.py
# author: Katherine Hegewisch (khegewisch@uidaho.edu)
#============================
# Required packages 
#============================
from netCDF4 import Dataset
import numpy as np
from datetime import date

#============================
# General Settings
#============================

# enter your bounding lat/lon for data extraction
latList = [41, 44, 41, 43.4]
lonList = [-100, -105, -110, -108]
NUM_POINTS = len(latList);

# enter your time start and time end
year_start = 1950
year_end = 2025
month_start=1
month_end=12

# enter variable name in the above opendap file
varname = 'tmin'  #units = deg C

# enter opendap filename
pathname = 'http://thredds.northwestknowledge.net:8080/thredds/dodsC/agg_terraclimate_'+varname+'_1950_CurrentYear_GLOBE.nc'
filehandle = Dataset(pathname,'r',format="NETCDF4")

#determine what the units are in the file
datahandle = filehandle.variables[varname]
#datahandle.variables   #look for units: deg_C

#determine what the scale and offset are for the variable
#datahandle.variables   #look for units: scale_factor: 0.01, add_offset: 0.0
scale_factor = datahandle.scale_factor
add_offset = datahandle.add_offset


#============================
# Data Subsetting 
#============================
# subset in time
timehandle=filehandle.variables['time']
time=timehandle[:]
time_min = (date(year_start,month_start,1)-date(1900,1,1)).days
time_max = (date(year_end,month_end,1)-date(1900,1,1)).days
time_index_min = (np.abs(time-time_min)).argmin()
time_index_max = (np.abs(time-time_max)).argmin()
time_index_range = range(time_index_min, time_index_max+1)
time = timehandle[time_index_range]


# subset in space (lat/lon)
lathandle = filehandle.variables['lat']
lonhandle = filehandle.variables['lon']
lat=lathandle[:]
lon=lonhandle[:]

myLat =[]
myLon =[]
myData =[]
for i in range(NUM_POINTS):
   # find indices of target lat/lon/day
   lat_index = (np.abs(lat-latList[i])).argmin()
   lon_index = (np.abs(lon-lonList[i])).argmin()

   #get grid centers extracted
   myLat.append(lat[lat_index])
   myLon.append(lon[lon_index])

   # subset data, applying the scale_factor and add_offset
   myData.append(add_offset + scale_factor *datahandle[time_index_range,lat_index,lon_index])

File: examples/point_subsets/terraclimate_extract_point_subsets.m
% example script for extracting point location data from terraclimate using OpenDap in MATLAB
% author: Katherine Hegewisch (khegewisch@uidaho.edu)
%===================================================
%enter your terraclimate variables
myVar = 'tmin';

%enter your terraclimate OPeNDAP URL
%find this here: http://thredds.northwestknowledge.net:8080/thredds/terraclimate_aggregated.html
myURL = ['http://thredds.northwestknowledge.net:8080/thredds/dodsC/agg_terraclimate_',myVar,'_1950_CurrentYear_GLOBE.nc'];

% enter your point locations as lat/lon pairs in the following arrays
latList=[41 44 45 41]; %list of north latitudes for each point location
lonList=[-110 -100 -101 -95]; %list of East longitudes for each point location
NUM_POINTS = length(latList);

%enter your time start and time end
year_start = 1950;
month_start=1;
year_end = 2025;
month_end=12;
%===================================================
timebounds_start =datenum(year_start,month_start,1)-datenum(1900,1,1);
timebounds_end = datenum(year_end,month_end,1)-datenum(1900,1,1);
timebounds = [timebounds_start timebounds_end];
lat=ncread(myURL,'lat');
lon=ncread(myURL,'lon');

time=ncread(myURL,'time');
[Y,M,D]=datevec(time+datenum(1900,1,1));
ftimei=find(time>=timebounds(1) & time<=timebounds(2));
time = time(ftimei);
NUM_TIME= length(time);

myData = NaN(NUM_POINTS,NUM_TIME);
myLat = NaN(NUM_POINTS,1);
myLon = NaN(NUM_POINTS,1);
for i= 1:NUM_POINTS;

        [dist,flati]=min(abs(lat-latList(i)));
        [dist,floni]=min(abs(lon-lonList(i)));
        myLat(i)=lat(flati);
        myLon(i)=lon(floni);

        % to change variables, see list here: http://thredds.northwestknowledge.net:8080/thredds/terraclimate_aggregated.html
        %dimensions of the data should be:
        myData(i,:)=ncread(myURL,myVar,[floni(1) flati(1) ftimei(1)],[length(floni) length(flati) length(ftimei)],[1 1 1]);
end;

%downloaded data is in myData (NUM_POINTS, NUM_TIME) with the grid cell centers myLat, myLon
